File: clients/go/transport.go
// How a request reaches the API, and what a server on the other end is not allowed to cost.
//
// The transport answers the status and the bytes and knows nothing of what the API declares:
// reading those bytes is the generated half's job, and deciding whether to send them again is the
// client's. That is what lets one HTTP implementation serve both the hand-written event path and
// every generated method — the generated package declares the shape it needs, and this type answers
// to it without either half importing the other.
//
// Nothing here reaches for a third-party HTTP library. Everything a server controls is bounded: how
// long one exchange may take, and how many bytes of body are read off the socket.

package hook0

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"runtime"
	"runtime/debug"
	"strings"
	"sync"
	"time"
)

const (
	// DefaultRequestTimeout is the longest one attempt at reaching the API is given before it is
	// abandoned.
	//
	// Ten seconds is far above what ingesting an event takes when the API is healthy, and short
	// enough that a stuck connection does not hold a caller for a noticeable time.
	DefaultRequestTimeout = 10 * time.Second

	// DefaultMaxResponseBytes is the largest response body read off a socket.
	DefaultMaxResponseBytes int64 = 8 * 1024 * 1024

	// MaxResponseHeaders is the most headers read out of one answer, and MaxHeaderBytes the longest
	// one of them may be.
	//
	// The head of an answer is written by the other end, so it is bounded like the body: a server
	// that is broken or hostile can otherwise spend a caller's memory on headers alone. Both are
	// the numbers the conformance corpus names, so that no two SDKs bound different things.
	MaxResponseHeaders = 64
	MaxHeaderBytes     = 64 * 1024

	// MaxHeadBytes is the largest whole head an answer may carry, every line counted together.
	//
	// This is the one that bounds what a head costs, because it bounds the total: a line count and
	// a size per line multiply, and the two above admit sixty-four lines of sixty-four kilobytes
	// between them. They earn their place by refusing early, on the line that crosses them rather
	// than at the end of the head; this one sets the ceiling.
	//
	// Sixteen kilobytes is the ceiling of the strictest runtime any target runs on, which is what
	// makes it a number every target can apply in library code. It is applied here rather than left
	// to MaxResponseHeaderBytes below: that one is an outer wall, set far above this so that what
	// refuses an abusive head is this client's own ceiling rather than whatever the runtime of the
	// day happens to allow.
	MaxHeadBytes = 16 * 1024
)

// jsonMediaType is what a request body says it carries, and what an answer is asked for in.
const jsonMediaType = "application/json"

// maxUserAgentPartChars is the longest each part this client composes its User-Agent out of may be,
// in characters.
//
// The runtime and the operating system are described by the platform rather than by this module, so
// their length is not this module's to guarantee: they are cut here so that the header cannot grow
// with whatever the platform feels like saying. Every part is also stripped of anything the grammar
// of the header uses as punctuation, so a platform cannot forge a shape it does not have.
const maxUserAgentPartChars = 64

// modulePath is what this module is imported under, which is the name the build records it by.
const modulePath = "github.com/hook0/hook0/clients/go"

// unknownVersion is what the version reads as when the build recorded none for this module, which
// is what a binary built without module information leaves behind.
const unknownVersion = "unknown"

// userAgent says which SDK, at which version, on which runtime and operating system, is talking to
// the API.
//
// Nothing in the module declares a version: a Go module is versioned by the tag that publishes it,
// so the number is read back out of what the build recorded rather than written down here, where it
// would disagree with that tag the first time either moved. Worked out once, since nothing it is
// built out of changes while the process runs.
var userAgent = sync.OnceValue(func() string {
	return fmt.Sprintf(
		"hook0-client-go/%s (%s; %s)",
		clipped(moduleVersion()),
		clipped(runtime.Version()),
		clipped(runtime.GOOS+" "+runtime.GOARCH),
	)
})

// moduleVersion is what the build recorded this module as: the version something else required it
// at, the version it was itself built as when it is the main module, and nothing nameable when the
// build recorded neither.
func moduleVersion() string {
	info, read := debug.ReadBuildInfo()
	if !read {
		return unknownVersion
	}
	for _, dependency := range info.Deps {
		if dependency.Path == modulePath && dependency.Version != "" {
			return dependency.Version
		}
	}
	if info.Main.Path == modulePath && info.Main.Version != "" {
		return info.Main.Version
	}
	return unknownVersion
}

// clipped is one part of the User-Agent, with everything the header's own grammar uses taken out of
// it and cut to maxUserAgentPartChars.
func clipped(part string) string {
	var kept strings.Builder
	for _, character := range part {
		if character < ' ' || character > '~' || character == '(' || character == ')' || character == ';' {
			continue
		}
		if kept.Len() == maxUserAgentPartChars {
			break
		}
		kept.WriteRune(character)
	}
	return kept.String()
}

// TransportError is a request that produced no answer to read.
//
// Several natures of failure land here — a connection that was refused or reset, an answer above a
// ceiling this client set for itself, a URL nothing can be sent to — and only the first of them
// could end differently. What a send retries is therefore decided by errors.Is against
// ErrUnreachable, ErrAnswerAboveABound and ErrUnusableAPIURL, never by this type: a client deciding
// by the type spends four attempts on a mistyped API URL and then hands its caller a message that
// accuses the network.
type TransportError struct {
	// Detail says what went wrong, in the words a caller is given.
	Detail string
	// Err is the nature of the failure, and under it whatever the standard library reported.
	Err error
}

// Error says why the API was not reached.
func (e *TransportError) Error() string {
	return e.Detail
}

// Unwrap answers what the standard library reported.
func (e *TransportError) Unwrap() error {
	return e.Err
}

// Transport issues one request and reads the answer.
//
// It answers the shape the generated package declares, so a generated operation group is built on
// one of these directly.
type Transport struct {
	baseURL          string
	token            string
	client           *http.Client
	maxResponseBytes int64
}

// NewTransport builds a transport reaching an instance of the API with a token valid for it.
//
// A timeout or a ceiling that names nothing is the default rather than no bound at all: a transport
// with no timeout is one a single hung connection holds forever.
func NewTransport(baseURL string, token string, timeout time.Duration, maxResponseBytes int64) *Transport {
	if timeout <= 0 {
		timeout = DefaultRequestTimeout
	}
	if maxResponseBytes <= 0 {
		maxResponseBytes = DefaultMaxResponseBytes
	}

	// The head is bounded where it is read off the socket rather than only after it arrived: a
	// server writing headers forever would otherwise be buffered whole before anything counted them.
	carrier := &http.Transport{}
	if standard, is := http.DefaultTransport.(*http.Transport); is {
		carrier = standard.Clone()
	}
	carrier.MaxResponseHeaderBytes = int64(MaxResponseHeaders) * int64(MaxHeaderBytes)

	return &Transport{
		baseURL:          baseURL,
		token:            token,
		client:           &http.Client{Timeout: timeout, Transport: carrier},
		maxResponseBytes: maxResponseBytes,
	}
}

// Request issues one request and answers the status, the body, and why it got neither.
//
// A refusal is an answer: the status and the body are what say whether repeating the request could
// end differently, so they are answered rather than raised over. Only a request that got no answer
// at all is an error here.
//
// This is the shape the generated package declares, which reads what the API sent and nothing about
// how it was sent. A caller that also needs the headers — the delay a paced instance names beside a
// refusal is one — asks Deliver for them.
func (t *Transport) Request(
	ctx context.Context,
	method string,
	path string,
	query url.Values,
	body any,
) (int, []byte, error) {
	status, _, payload, err := t.Deliver(ctx, method, path, query, body)
	return status, payload, err
}

// Deliver issues one request and answers the status, the headers, and the body.
//
// It is Request with what the answer carried beside its body, which is what a client reads when the
// API names how long to wait before the request becomes servable again.
func (t *Transport) Deliver(
	ctx context.Context,
	method string,
	path string,
	query url.Values,
	body any,
) (int, http.Header, []byte, error) {
	target, err := t.resolve(path, query)
	if err != nil {
		return 0, nil, nil, &TransportError{
			Detail: err.Error(),
			Err:    fmt.Errorf("%w: %w", ErrUnusableAPIURL, err),
		}
	}

	var encoded io.Reader
	if body != nil {
		written, err := json.Marshal(body)
		if err != nil {
			// A body that cannot be written is not a request that could be repeated.
			return 0, nil, nil, fmt.Errorf("the request body cannot be written as JSON: %w", err)
		}
		encoded = bytes.NewReader(written)
	}

	request, err := http.NewRequestWithContext(ctx, method, target, encoded)
	if err != nil {
		return 0, nil, nil, fmt.Errorf("the request cannot be built: %w", err)
	}
	request.Header.Set("Authorization", "Bearer "+t.token)
	request.Header.Set("Accept", jsonMediaType)
	// Set rather than left alone: what the standard library names itself here says which Go built
	// the caller and nothing at all about which SDK is talking.
	request.Header.Set("User-Agent", userAgent())
	if body != nil {
		request.Header.Set("Content-Type", jsonMediaType)
	}

	answer, err := t.client.Do(request)
	if err != nil {
		return 0, nil, nil, &TransportError{
			Detail: err.Error(),
			Err:    fmt.Errorf("%w: %w", ErrUnreachable, err),
		}
	}
	defer answer.Body.Close()

	if err := bounded(answer.Header); err != nil {
		return 0, nil, nil, err
	}

	payload, err := io.ReadAll(io.LimitReader(answer.Body, t.maxResponseBytes+1))
	if err != nil {
		// A body that stopped mid-way is a request that got no answer worth reading, and the next
		// one could carry the whole of it.
		return 0, nil, nil, &TransportError{
			Detail: err.Error(),
			Err:    fmt.Errorf("%w: %w", ErrUnreachable, err),
		}
	}
	if int64(len(payload)) > t.maxResponseBytes {
		return 0, nil, nil, &TransportError{
			Detail: fmt.Sprintf("the API answered more than the %d bytes read at most", t.maxResponseBytes),
			Err:    ErrAnswerAboveABound,
		}
	}

	return answer.StatusCode, answer.Header, payload, nil
}

// bounded refuses a head above what this client agrees to hold, naming the ceiling it crossed.
func bounded(headers http.Header) error {
	if len(headers) > MaxResponseHeaders {
		return &TransportError{
			Detail: fmt.Sprintf(
				"the API answered %d headers, above the %d read at most", len(headers), MaxResponseHeaders,
			),
			Err: ErrAnswerAboveABound,
		}
	}

	whole := 0
	for name, values := range headers {
		for _, value := range values {
			if len(name)+len(value) > MaxHeaderBytes {
				return &TransportError{
					Detail: fmt.Sprintf(
						"the API answered a `%s` header above the %d bytes read at most", name, MaxHeaderBytes,
					),
					Err: ErrAnswerAboveABound,
				}
			}
			whole += len(name) + len(value)
		}
	}

	// The total, once every line has been counted. A head this far along is one the runtime has
	// already read whole, so counting it to its end costs nothing more than counting part of it,
	// and the refusal can say how heavy the head actually was.
	if whole > MaxHeadBytes {
		return &TransportError{
			Detail: fmt.Sprintf(
				"the API answered a head of %d bytes, above the %d read at most", whole, MaxHeadBytes,
			),
			Err: ErrAnswerAboveABound,
		}
	}

	return nil
}

// resolve says where a request lands: a path of its own replaces the base's, a relative one extends
// it, and the query the caller assembled is added to whatever the base already carried.
func (t *Transport) resolve(path string, query url.Values) (string, error) {
	base, err := url.Parse(t.baseURL)
	if err != nil {
		return "", fmt.Errorf("`%s` is not an API URL: %w", t.baseURL, err)
	}
	if !strings.HasSuffix(base.Path, "/") {
		base.Path += "/"
	}

	reference, err := url.Parse(path)
	if err != nil {
		return "", fmt.Errorf("`%s` is not a path: %w", path, err)
	}

	resolved := base.ResolveReference(reference)
	if len(query) > 0 {
		carried := resolved.Query()
		for name, values := range query {
			for _, value := range values {
				carried.Add(name, value)
			}
		}
		resolved.RawQuery = carried.Encode()
	}

	return resolved.String(), nil
}

File: clients/php/src/Transport.php
<?php

declare(strict_types=1);

namespace Hook0;

/**
 * How a request reaches the API, and what a server on the other end is not allowed to cost.
 *
 * The transport answers the status and the bytes and knows nothing of what the API declares: reading
 * those bytes is the generated half's job, and deciding whether to send them again is the client's.
 * That is what lets one HTTP implementation serve both the hand-written event path and every
 * generated method — a generated group calls whatever object it is handed, and this is the one this
 * package ships.
 *
 * Nothing here reaches for a third-party HTTP library. What the bundled one bounds on its own is
 * little: it refuses to follow a redirect unless asked, and it holds one response header line to a
 * hundred kilobytes. It bounds neither how many header lines it accepts, nor the head as a whole,
 * nor the body, and its own per-line ceiling sits above the one the shared corpus names. So every
 * ceiling that matters is applied here, in the callbacks the exchange is read through, which is also
 * what lets a head past the count or a line past its length be refused before the body is read at
 * all.
 */
final class Transport
{
    /**
     * Longest one attempt at reaching the API is given before it is abandoned, in seconds.
     *
     * Ten seconds is far above what ingesting an event takes when the API is healthy, and short
     * enough that a stuck connection does not hold a caller for a noticeable time.
     */
    public const DEFAULT_REQUEST_TIMEOUT = 10.0;

    /** Largest response body read off a socket, in bytes. */
    public const DEFAULT_MAX_RESPONSE_BYTES = 8 * 1024 * 1024;

    /**
     * How many header lines an answer may carry before it is refused.
     *
     * Sixty-four is well above what the API sends, and the point of the bound is to refuse early:
     * the line that crosses it stops the exchange before the body is read.
     */
    public const DEFAULT_MAX_RESPONSE_HEADERS = 64;

    /** Longest one header line may be, name and value together, in bytes. */
    public const DEFAULT_MAX_HEADER_BYTES = 64 * 1024;

    /**
     * Largest whole head an answer may carry, every line counted together, in bytes.
     *
     * This is the one that bounds what a head costs. A line count and a size per line multiply:
     * sixty-four lines of sixty-four kilobytes each is four megabytes of head, and both of the
     * bounds above admit it. They earn their place by refusing early, on the line that crosses them
     * rather than at the end of the head; this one sets the ceiling.
     */
    public const DEFAULT_MAX_HEAD_BYTES = 16 * 1024;

    /** What a request body says it carries, and what an answer is asked for in. */
    private const JSON_MEDIA_TYPE = 'application/json';

    /**
     * Longest each part this client composes its `User-Agent` out of may be, in characters.
     *
     * The runtime and the operating system are described by the platform rather than by this package,
     * so their length is not this package's to guarantee: they are cut here so that the header cannot
     * grow with whatever the platform feels like saying. Every part is also stripped of anything the
     * grammar of the header uses as punctuation, so a platform cannot forge a shape it does not have.
     */
    private const MAX_USER_AGENT_PART_CHARS = 64;

    /** What this package is installed under, which is what its version is asked for by. */
    private const PACKAGE_NAME = 'hook0/client';

    /** What a version reads as where nothing that knows one is there to be asked. */
    private const UNKNOWN_VERSION = 'unknown';

    /** The schemes this transport reaches. */
    private const SCHEMES = ['http', 'https'];

    /** What a URL carrying a scheme of its own looks like. */
    private const ABSOLUTE_URL = '#^[A-Za-z][A-Za-z0-9+.\-]*://#';

    /** Shortest timeout the underlying library reads as a timeout rather than as none at all. */
    private const MIN_TIMEOUT_MS = 1;

    /**
     * @param string $baseUrl where the API lives, such as https://app.hook0.com/api/v1
     * @param string $token an authentication token valid for that API
     * @param float $timeout how long one attempt is given, in seconds
     * @param int $maxResponseBytes the largest answer read off a socket
     * @param int $maxResponseHeaders how many header lines an answer may carry
     * @param int $maxHeaderBytes the longest one header line may be
     * @param int $maxHeadBytes the largest whole head, every line counted together
     */
    public function __construct(
        private readonly string $baseUrl,
        private readonly string $token,
        private readonly float $timeout = self::DEFAULT_REQUEST_TIMEOUT,
        private readonly int $maxResponseBytes = self::DEFAULT_MAX_RESPONSE_BYTES,
        private readonly int $maxResponseHeaders = self::DEFAULT_MAX_RESPONSE_HEADERS,
        private readonly int $maxHeaderBytes = self::DEFAULT_MAX_HEADER_BYTES,
        private readonly int $maxHeadBytes = self::DEFAULT_MAX_HEAD_BYTES,
    ) {
    }

    /**
     * What the API answered, whether or not it answered a success.
     *
     * This is the shape the generated half of this package reads, which is the status and the bytes.
     * A caller that also needs what the answer carried beside its body — the delay a paced instance
     * names is one — asks {@see self::deliver()} for it.
     *
     * @param list<array{0: string, 1: string}> $query the name and value pairs of the query string
     * @param mixed $body what to send as a JSON document, or nothing at all
     * @return array{0: int, 1: string} the status and the body
     */
    public function request(string $method, string $path, array $query = [], mixed $body = null): array
    {
        [$status, , $payload] = $this->deliver($method, $path, $query, $body);

        return [$status, $payload];
    }

    /**
     * What the API answered, headers included, whether or not it answered a success.
     *
     * Header names are lowercased and a later value wins over an earlier one under the same name, so
     * a caller reads a header without knowing which case the server wrote it in.
     *
     * @param list<array{0: string, 1: string}> $query the name and value pairs of the query string
     * @param mixed $body what to send as a JSON document, or nothing at all
     * @return array{0: int, 1: array<string, string>, 2: string} the status, the headers and the body
     */
    public function deliver(string $method, string $path, array $query = [], mixed $body = null): array
    {
        return $this->exchange($method, $this->resolved($path, $query), $body);
    }

    /**
     * Where a request lands: a path of its own replaces the base's, a relative one extends it.
     */
    private function resolved(string $path, array $query): string
    {
        $target = $this->joined($path);

        $parts = parse_url($target);
        if ($parts === false || !isset($parts['scheme'], $parts['host'])) {
            throw TransportError::unusableApiUrl(
                sprintf('`%s` is not somewhere this transport can send a request', $target)
            );
        }
        if (!in_array(strtolower($parts['scheme']), self::SCHEMES, true) || $parts['host'] === '') {
            throw TransportError::unusableApiUrl(
                sprintf('`%s` is not somewhere this transport can send a request', $target)
            );
        }
        if (count($query) === 0) {
            return $target;
        }

        $written = [];
        foreach ($query as [$name, $value]) {
            $written[] = rawurlencode($name) . '=' . rawurlencode($value);
        }
        $separator = str_contains($target, '?') ? '&' : '?';

        return $target . $separator . implode('&', $written);
    }

    /** The base URL and the path an operation names, put together the way a browser would. */
    private function joined(string $path): string
    {
        if (preg_match(self::ABSOLUTE_URL, $path) === 1) {
            return $path;
        }

        $base = rtrim($this->baseUrl, '/') . '/';
        if (!str_starts_with($path, '/')) {
            return $base . $path;
        }

        $parts = parse_url($base);
        if ($parts === false || !isset($parts['scheme'], $parts['host'])) {
            throw TransportError::unusableApiUrl(
                sprintf('`%s` is not a base URL a path can be resolved against', $this->baseUrl)
            );
        }
        $port = isset($parts['port']) ? ':' . $parts['port'] : '';

        return $parts['scheme'] . '://' . $parts['host'] . $port . $path;
    }

    /**
     * One exchange, bounded on every axis a server controls.
     *
     * @return array{0: int, 1: array<string, string>, 2: string}
     */
    private function exchange(string $method, string $target, mixed $body): array
    {
        $headers = [];
        $payload = '';
        $head = 0;
        $lines = 0;
        $refused = null;

        $handle = curl_init();
        curl_setopt_array($handle, $this->options($method, $target, $body));
        curl_setopt(
            $handle,
            CURLOPT_HEADERFUNCTION,
            function (\CurlHandle $handle, string $line) use (&$headers, &$head, &$lines, &$refused): int {
                $read = strlen($line);
                $refused ??= $this->refusedHead($line, $head + $read, $lines + 1);
                if ($refused !== null) {
                    return 0;
                }

                $head += $read;
                $name = strstr($line, ':', true);
                if ($name !== false) {
                    $lines++;
                    $headers[strtolower(trim($name))] = trim(substr($line, strlen($name) + 1));
                }

                return $read;
            }
        );
        curl_setopt(
            $handle,
            CURLOPT_WRITEFUNCTION,
            function (\CurlHandle $connection, string $chunk) use (&$payload, &$refused): int {
                if (strlen($payload) + strlen($chunk) > $this->maxResponseBytes) {
                    $refused ??= sprintf(
                        'the API answered more than the %d bytes read at most',
                        $this->maxResponseBytes
                    );

                    return 0;
                }
                $payload .= $chunk;

                return strlen($chunk);
            }
        );

        $answered = curl_exec($handle);
        $failure = curl_error($handle);
        $status = (int) curl_getinfo($handle, CURLINFO_RESPONSE_CODE);
        unset($handle);

        if ($refused !== null) {
            throw TransportError::answerAboveABound($refused);
        }
        if ($answered === false) {
            throw TransportError::noAnswer($failure === '' ? 'the API answered nothing' : $failure);
        }

        return [$status, $headers, $payload];
    }

    /**
     * Why a head this long, or this many lines in, is one this client will not read, when it is.
     *
     * The status line carries no name and is not counted as a header, but its bytes are: it is part
     * of what the head costs.
     */
    private function refusedHead(string $line, int $head, int $lines): ?string
    {
        $named = strstr($line, ':', true) !== false;

        if ($named && strlen(rtrim($line, "\r\n")) > $this->maxHeaderBytes) {
            $name = strtolower(trim((string) strstr($line, ':', true)));

            return sprintf(
                'the API answered a `%s` header above the %d bytes read at most',
                $name,
                $this->maxHeaderBytes
            );
        }
        if ($named && $lines > $this->maxResponseHeaders) {
            return sprintf(
                'the API answered more than the %d header lines read at most',
                $this->maxResponseHeaders
            );
        }
        if ($head > $this->maxHeadBytes) {
            return sprintf(
                'the API answered a head above the %d bytes read at most',
                $this->maxHeadBytes
            );
        }

        return null;
    }

    /**
     * Everything one request is issued under, beyond the callbacks that bound what it may cost.
     *
     * @return array<int, mixed>
     */
    private function options(string $method, string $target, mixed $body): array
    {
        $milliseconds = max(self::MIN_TIMEOUT_MS, (int) round($this->timeout * 1000));

        $lines = [
            'Authorization: Bearer ' . $this->token,
            'Accept: ' . self::JSON_MEDIA_TYPE,
            'User-Agent: ' . self::userAgent(),
            // The library asks the server for permission before sending a body of any size, and
            // waits a second for an answer that a plain HTTP/1.1 server never sends. Nothing here
            // needs that handshake, so it is turned off rather than waited out.
            'Expect:',
        ];

        $options = [
            CURLOPT_URL => $target,
            CURLOPT_CUSTOMREQUEST => strtoupper($method),
            CURLOPT_RETURNTRANSFER => true,
            // A redirect is an answer, not a place to follow: chasing one would send the credential
            // to wherever the answer pointed, as many times as it kept pointing.
            CURLOPT_FOLLOWLOCATION => false,
            CURLOPT_TIMEOUT_MS => $milliseconds,
            CURLOPT_CONNECTTIMEOUT_MS => $milliseconds,
            // Timeouts under a second are only honoured when the library is told not to reach for
            // the alarm signal it otherwise uses to interrupt name resolution.
            CURLOPT_NOSIGNAL => true,
        ];

        if ($body !== null) {
            $lines[] = 'Content-Type: ' . self::JSON_MEDIA_TYPE;
            $options[CURLOPT_POSTFIELDS] = Runtime::encode($body);
        }
        $options[CURLOPT_HTTPHEADER] = $lines;

        return $options;
    }

    /**
     * Which SDK, at which version, on which runtime and operating system, is talking to the API.
     *
     * A Composer package carries no version of its own: the registry reads one off the tag that
     * publishes it, which is why `composer.json` declares none and whatever installed this package is
     * the only thing that can be asked. A source checkout that went through no installer has nothing
     * to ask, and says so rather than naming a version it cannot stand behind. So does a project that
     * has Composer but copied these sources in rather than requiring them: asking after a package
     * nothing installed raises rather than answering, and a header is no place to find that out.
     */
    private static function userAgent(): string
    {
        $installed = class_exists(\Composer\InstalledVersions::class)
            && \Composer\InstalledVersions::isInstalled(self::PACKAGE_NAME)
            ? \Composer\InstalledVersions::getPrettyVersion(self::PACKAGE_NAME)
            : null;

        return sprintf(
            'hook0-client-php/%s (%s; %s)',
            self::clipped($installed ?? self::UNKNOWN_VERSION),
            self::clipped('php ' . PHP_VERSION),
            self::clipped(PHP_OS_FAMILY . ' ' . php_uname('m'))
        );
    }

    /**
     * One part of the `User-Agent`, with everything the header's own grammar uses taken out of it and
     * cut to {@see self::MAX_USER_AGENT_PART_CHARS}.
     */
    private static function clipped(string $part): string
    {
        $printable = (string) preg_replace('/[^\x20-\x7E]/', '', $part);

        return substr(str_replace(['(', ')', ';'], '', $printable), 0, self::MAX_USER_AGENT_PART_CHARS);
    }
}
